### PR TITLE
Replace costly sorting of VDIF frames by iterating over dict.

### PR DESCRIPTION
**Pull request by mhvk**
@cczhu - I tried to speed up the frame set code, though it ends up helping less than I had thought based on your sample code. So, not sure whether or not this is worth merging.

Somewhat relatedly, I wonder a bit about the use of the sort keyword...  At least when one passes in `thread_ids`, it seems one always should sort.

### CODE REVIEW

**Comment by cczhu** [comment]
I ran a few profiling tests, and it indeed looks like using the OrderedDict takes slightly *longer* than not doing so (5 - 6 seconds over a 2e9 sample decode of the B1957 EVN run).  While the final sorting is faster, apparently the extra time is taken up by creating an empty OrderedDict and populating it with frames (which is strange since the keys don't get sorted during that process).  It might not be a good idea to merge.

I do agree that `sort = False` really shouldn't ever be used.  I can get rid of it in a quick PR.

**Comment by mhvk** [comment]
I think the profiling may be python-version dependent, since `OrderedDict` became implemented in C only recently. In any case, let's give up on this one - it really doesn't make things better/clearer/faster - and let's indeed get rid of `sort` (I think it may have been there before one could pass in specific threads.)

**Comment by cczhu** [comment]
On a second inspection of my timing results, I realize I made a mistake - in my "fast sequential code" (the one that bypasses the stream reader), using OrderedDict makes the code take 5 seconds longer, but using the stream reader to do sequential decoding, using OrderedDict makes the code take about 6 seconds less.  So we do see some advantage to using OrderedDict in more realistic run environments, though I'm honestly not sure why it slows my fast code.

More interestingly, I read on [Reddit](https://www.reddit.com/r/Python/comments/51te8x/raymond_hettingerpython36_news_ordereddict_is/) that `dict` is now superior to `OrderedDict` for sorting.  I tried a version of your code that  uses dict, and now get a 5 second boost in my fast sequential code, and a 16 second boost when using the stream reader.  (When not using the line profiler, these gains are halved.)  Since the overall code takes ~550 s to run with the line profiler (~250 s when just using `time python`), this is a very small gain, but maybe we should resurrect this PR?

**Comment by mhvk** [comment]
Indeed, in python >=3.6, `dict` itself is ordered by default, but we cannot count on this. Anyway, I think there is enough reason to think a little more about this, so restored the branch and reopened the PR.

**Comment by cczhu** [comment]
That was my understanding too, but the all the tests I did were on a Python 2.7 environment.  I wonder if this is somehow Scinet-specific, or OrderedDict is just always slower?

**Comment by mhvk** [comment]
`OrderedDict` is slower on `python2` than on more recent versions (it is now implemented in C).

**Comment by cczhu** [comment]
I tried this little test on my own workstation (Py 2.7 and 3.5) and Scinet (Py 2.7):
```
from collections import OrderedDict
import numpy as np
from baseband import vdif
from baseband.data import SAMPLE_VDIF

# Get frame.
fh = vdif.open(SAMPLE_VDIF, 'rs')
fr = fh._frameset.frames[0]

# Get threads (max number of threads is 1024)
thread_ids = np.arange(1024)
np.random.shuffle(thread_ids)

# Populate fake frameset
frameset = []
for i in thread_ids:
    header = fr.header.copy()
    header['thread_id'] = i
    frameset.append(vdif.frame.VDIFFrame(header, vdif.payload.VDIFPayload(fr.payload.words.copy(), header=header)))
zipset = zip(thread_ids, frameset)

# Functions for populating dicts or ordered dicts
def make_ordered_dict():
    mydict = OrderedDict()
    for i, item in enumerate(thread_ids):
        mydict[item] = frameset[i]
    return mydict

def make_dict():
    mydict = {}
    for i, item in enumerate(thread_ids):
        mydict[item] = frameset[i]
    return mydict

# Test sorting list
%timeit -r 1 -n 1 frameset.sort(key=lambda frame: frame['thread_id'])

# Test sorting ordered dict
%timeit -r 1 -n 1 frameordered = make_ordered_dict(); sortedkeys = sorted(frameordered.keys()); frames = [frameordered[thread_id] for thread_id in sortedkeys]

# Test sorting dict
%timeit -r 1 -n 1 framedict = make_dict(); sortedkeys = sorted(framedict.keys()); frames = [framedict[thread_id] for thread_id in sortedkeys]
```
The results consistently show that list sorting is the slowest at >1.5 ms, creating and sorting with the ordered dict takes about half that time (or more for Py 2.7), and creating and sorting with the dict takes a fifth or a tenth of the time.

**Comment by mhvk** [comment]
@cczhu - if we remove the option *not* to sort, we don't need the `OrderedDict`, and the code becomes quite a bit cleaner. I think it may now be worth it...

**Comment by mhvk** [comment]
And given your comment just now, this will also be the fastest solution!

File: baseband/vdif/frame.py [hunk 1]
```diff
@@ -228,19 +221,21 @@ def fromfile(cls, fh, thread_ids=None, sort=True, edv=None, verify=True):
         Returns
         -------
         frameset : VDIFFrameSet instance
-            Holds ''frames'' property with a possibly sorted list of frames.
+            Holds ''frames'' property a sorted list of frames (either by
```
Comment (cczhu) [inline baseband/vdif/frame.py]: Reword first sentence to: 
```
Its ``frames`` property holds a sorted list of frames (either by ``thread_id`` or following the input ``thread_ids`` list). 
```

File: baseband/vdif/frame.py [hunk 2]
```diff
@@ -257,13 +252,11 @@ def fromfile(cls, fh, thread_ids=None, sort=True, edv=None, verify=True):
...
-            if thread_ids:
-                frames.sort(key=lambda frame:
-                            thread_ids.index(frame['thread_id']))
-            else:
-                frames.sort(key=lambda frame: frame['thread_id'])
+        # Turn dict of frames into a list, following order given by
+        # thread_ids, or just sorting by their own thread_id
+        if thread_ids is None:
+            thread_ids = sorted(frames.keys())
+        frames = [frames[thread_id] for thread_id in thread_ids]
```
Comment (cczhu) [inline baseband/vdif/frame.py]: The linter in my text editor complains that `thread_id` is modified by this line.  Not sure if it's best practice to create a unique dummy variable for list comprehension.

**Comment by mhvk** [comment]
OK, fixed those two little things (not sure about the IDE thing - `thread_id` was not used after - but changed it anyway). Will merge now.

**Pull request merged**